File: CPP/06 Plugins/vehicle_plugins/src/vehicle_plugins.cpp
/**
 * @file vehicle_plugins.cpp
 *
 * @brief Definition of the derived class that will be used as plugin
 *
 * @author Antonio Mauro Galiano
 * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
 *
 */


#include "vehicle_base/regular_vehicle.hpp"

// derived classes declaration which will be instanciated through pluginlib
namespace vehicle_plugins
{
  class Motorbike : public vehicle_base::RegularVehicle
  {
    public:
      void initialize(std::string vehicleType) override
      {
        vehicleType_ = vehicleType;
      }

      int setNumPassengers() override
      {
        return 2;
      }

      std::string getVehicleType() override
      {
        return vehicleType_;
      }

    protected:
      std::string vehicleType_;
  };

  class Bicycle : public vehicle_base::RegularVehicle
  {
    public:
      void initialize(std::string vehicleType) override
      {
        vehicleType_ = vehicleType;
      }

      int setNumPassengers() override
      {
        return 1;
      }

      std::string getVehicleType() override
      {
        return vehicleType_;
      }

    protected:
      std::string vehicleType_;
  };

  class Truck : public vehicle_base::RegularVehicle
  {
    public:
      void initialize(std::string vehicleType) override
      {
        vehicleType_ = vehicleType;
      }

      int setNumPassengers() override
      {
        return 5;
      }

      std::string getVehicleType() override
      {
        return vehicleType_;
      }

    protected:
      std::string vehicleType_;
  };
}

#include <pluginlib/class_list_macros.hpp>

// registration of the derived classes as plugin
PLUGINLIB_EXPORT_CLASS(vehicle_plugins::Motorbike, vehicle_base::RegularVehicle)
PLUGINLIB_EXPORT_CLASS(vehicle_plugins::Bicycle, vehicle_base::RegularVehicle)
PLUGINLIB_EXPORT_CLASS(vehicle_plugins::Truck, vehicle_base::RegularVehicle)

File: PYTHON/plugins/vehicle_plugins_py/vehicle_plugins_py/vehicle_plugins.py
#  *
#  * @file vehicle_plugins.py
#  *
#  * @brief Definition of the derived classes that will be used as plugins
#  *
#  * @author Antonio Mauro Galiano
#  * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
#  *
#

from vehicle_base_py import RegularVehicle


class Motorbike(RegularVehicle):
    """Motorbike plugin implementation."""

    def __init__(self):
        self.vehicle_type_ = None

    def initialize(self, vehicle_type: str) -> None:
        """Initialize the motorbike."""
        self.vehicle_type_ = vehicle_type

    def set_num_passengers(self) -> int:
        """Return the number of passengers for a motorbike."""
        return 2

    def get_vehicle_type(self) -> str:
        """Return the vehicle type."""
        return self.vehicle_type_


class Bicycle(RegularVehicle):
    """Bicycle plugin implementation."""

    def __init__(self):
        self.vehicle_type_ = None

    def initialize(self, vehicle_type: str) -> None:
        """Initialize the bicycle."""
        self.vehicle_type_ = vehicle_type

    def set_num_passengers(self) -> int:
        """Return the number of passengers for a bicycle."""
        return 1

    def get_vehicle_type(self) -> str:
        """Return the vehicle type."""
        return self.vehicle_type_


class Truck(RegularVehicle):
    """Truck plugin implementation."""

    def __init__(self):
        self.vehicle_type_ = None

    def initialize(self, vehicle_type: str) -> None:
        """Initialize the truck."""
        self.vehicle_type_ = vehicle_type

    def set_num_passengers(self) -> int:
        """Return the number of passengers for a truck."""
        return 5

    def get_vehicle_type(self) -> str:
        """Return the vehicle type."""
        return self.vehicle_type_


# Entry points for plugin discovery
# These functions will be registered as entry points in setup.py
def create_motorbike():
    """Factory function to create a Motorbike instance."""
    return Motorbike()


def create_bicycle():
    """Factory function to create a Bicycle instance."""
    return Bicycle()


def create_truck():
    """Factory function to create a Truck instance."""
    return Truck()

